### PR TITLE
WIP: invoke first pass

### PR DESCRIPTION
**Pull request by zootalures**

### CODE REVIEW

File: api/agent/protocol/json.go [hunk 1]
```diff
@@ -37,132 +46,154 @@ type CallResponseHTTP struct {
...
+	err = json.NewEncoder(h.in).Encode(in)
+	if err != nil {
+		return "", err
+	}
+	return callID, nil
+
 }
 
-func (h *JSONProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) error {
+func (h *JSONProtocol) Dispatch(ctx context.Context, ci *event.Event) (*event.Event, error) {
```
Comment (rdallman) [inline api/agent/protocol/json.go]: eventually I think this is probably the right thing to do (actually, kill dispatch is probably the right thing to do, the contract is whack imo even if the abstraction was useful while we needed it), and now that the code is all here we should probably proceed, I guess this seems like a lot of extra work to add here and test considering we're going to kill http/json in the relatively near future. am not sure it's worth undoing, the dispatch signature change does seem responsible for a large number of lines in this patch and I do think we'll kill it. previously I guess we were only writing back an http request, we could likely following dispatch translate that into an event (if it's not already), which would mean that there's only one code path that needs to do that translation (vs 2). again, not sure it's worth undoing, thinking out loud.

**Comment by rdallman** [comment]
i'll try to spend a few more hours on this this weekend, sorry this took so long... caught up on my email and in the clear now, realize my review is vague, i'm musing about feasible reductions (and have not yet tried to run / fixup tests with this, maybe just press on).

**Comment by zootalures** [comment]
Thanks - really appreciate it  - I think i bit of a bit more than my time allowed here

**Comment by skinowski** [comment]
I know this is WIP, but glanced through it and there are too many seemingly not 1-1 related changes stacked: renaming agent config, removing max-req-size, adding events & documents, reworking IO in various parts, modifying error.go, etc.

**Comment by rdallman** [comment]
> renaming agent config, removing max-req-size

yeah, good idea, i'll try to post a patch with just these today, easy enough to extract those two out.

**Comment by skinowski** [comment]
I wonder if we can merge the IO changes in advance also. By accepting and incurring an extra buffer/copy, we can perhaps merge the IO changes (pure-runner, runner-client, etc.) and get them tested/merged before invoke/events changes make it.

**Comment by rdallman** [comment]
ok #1144 and #1145 are up... i think we can probably stage the changes that rip http.request out of call out too without any of the i/o changes, which adds context.Context to the agent methods and updates the tests and such, that should reduce the noise a good bit more from here after another rebase... i'll give it a couple hours and see if it's worth it

**Comment by rdallman** [comment]
as promised, have been futzing with the pure runner to get system tests to pass. this patch does a lot of changes there that I guess I was trying to avoid, I haven't quite made up my mind if we can get away with it or not i'm still trying to get the system tests to pass. i'll try to press on tomorrow

**Comment by riconnon** [comment]
I rebased this so we can see the remaining diff...
obviously it's still huge and we may or may not go with it as is

File: api/agent/agent.go [hunk 2]
```diff
@@ -71,15 +73,15 @@ import (
...
+	GetCall(ctx context.Context, opts ...CallOpt) (Call, error)
 
 	// Submit will attempt to execute a call locally, a Call may store information
 	// about itself in its Start and End methods, which will be called in Submit
 	// immediately before and after the Call is executed, respectively. An error
 	// will be returned if there is an issue executing the call or the error
 	// may be from the call's execution itself (if, say, the container dies,
 	// or the call times out).
-	Submit(Call) error
+	Submit(ctx context.Context, call Call) (*event.Event, error)
```
[the remark this answers is not in the document]
Comment (zootalures) [reply api/agent/agent.go]: yep, definitely

**Comment by rdallman** [comment]
@zootalures @riconnon @mantree let's close this? it's here and we want to pull some of this over in some capacity but since we have a slightly different first pass in now... ? feel free to re-open

**Pull request closed without merging**